import java.util.*;

class Solution {
    public List<List<Integer>> palindromePairs(String[] words) {
        List<List<Integer>> pairs = new ArrayList<>();
        Map<String, Integer> map = new HashMap<>();
        
        for (int i = 0; i < words.length; i++) {
            map.put(words[i], i);
        }
        
        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            
            for (int j = 0; j <= word.length(); j++) {
                String str1 = word.substring(0, j);
                String str2 = word.substring(j);
                
                if (isPalindrome(str1)) {
                    String revStr2 = new StringBuilder(str2).reverse().toString();
                    if (map.containsKey(revStr2) && map.get(revStr2) != i) {
                        pairs.add(Arrays.asList(map.get(revStr2), i));
                    }
                }
                
                if (isPalindrome(str2) && str2.length() > 0) {
                    String revStr1 = new StringBuilder(str1).reverse().toString();
                    if (map.containsKey(revStr1) && map.get(revStr1) != i) {
                        pairs.add(Arrays.asList(i, map.get(revStr1)));
                    }
                }
            }
        }
        
        return pairs;
    }
    
    private boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left++) != s.charAt(right--)) {
                return false;
            }
        }
        return true;
    }
}
